src/cmd/builtin/fsck.c

```diff
@@ -0,0 +1,266 @@
+#include "src/common/libkvs/treeobj.h"
+#include "src/common/libkvs/kvs_checkpoint.h"
+#include "src/common/libutil/fsd.h"
+#include "src/common/libutil/blobref.h"
+#include "src/common/libcontent/content.h"
+#include "ccan/str/str.h"
+
+#include "builtin.h"
+
+static void fsck_treeobj (flux_t *h,
+                          const char *path,
+                          json_t *treeobj);
+
+static bool verbose;
+static int errorcount;
+
+static void read_verror (const char *fmt, va_list ap)
+{
+    char buf[128];
+    vsnprintf (buf, sizeof (buf), fmt, ap);
+    fprintf (stderr, "%s\n", buf);
+}
+
+static __attribute__ ((format (printf, 1, 2)))
+void read_error (const char *fmt, ...)
+{
+    va_list ap;
+    va_start (ap, fmt);
+    read_verror (fmt, ap);
+    va_end (ap);
+}
+
+static void fsck_valref (flux_t *h,
+                         const char *path,
+                         json_t *treeobj)
+{
+    int count = treeobj_get_count (treeobj);
+    const void *buf;
+    size_t buflen;
+
+    for (int i = 0; i < count; i++) {
+        flux_future_t *f;
+        if (!(f = content_load_byblobref (h,
+                                          treeobj_get_blobref (treeobj, i),
+                                          CONTENT_FLAG_CACHE_BYPASS))
+            || content_load_get (f, &buf, &buflen) < 0) {
+            if (errno == ENOENT)
+                read_error ("%s: missing blobref index=%d",
+                            path,
+                            i);
+            else
+                read_error ("%s: error retrieving blobref index=%d: %s",
+                            path,
+                            i,
+                            future_strerror (f, errno));
+            errorcount++;
+            flux_future_destroy (f);
+            return;
+        }
+        flux_future_destroy (f);
+    }
+}
+
+static void fsck_val (flux_t *h,
+                      const char *path,
+                      json_t *treeobj)
+{
+    /* Do nothing for now */
+}
+
+static void fsck_symlink (flux_t *h,
+                          const char *path,
+                          json_t *treeobj)
+{
+    /* Do nothing for now */
+}
+
+static void fsck_dir (flux_t *h,
+                      const char *path,
+                      json_t *treeobj)
+{
+    json_t *dict = treeobj_get_data (treeobj);
+    const char *name;
+    json_t *entry;
+
+    json_object_foreach (dict, name, entry) {
+        char *newpath;
+        if (asprintf (&newpath, "%s.%s", path, name) < 0)
+            log_msg_exit ("out of memory");
+        fsck_treeobj (h, newpath, entry); // recurse
+        free (newpath);
+    }
+}
+
+static void fsck_dirref (flux_t *h,
+                         const char *path,
+                         json_t *treeobj)
+{
+    flux_future_t *f;
+    const void *buf;
+    size_t buflen;
+    json_t *treeobj_deref = NULL;
+
+    if (treeobj_get_count (treeobj) != 1)
+        log_msg_exit ("%s: blobref count is not 1", path);
```

Comment:
Should this be fatal?  Maybe it should be a `read_error()` and say something like
```
%s: dirref object has blob count of %d, when 1 is expected
```
(and return rather than trying to recurse)

Reply:
I did think about this as I wrote the tool.  Since we wouldn't have the ability to recurse, my mind meandered towards "I guess we stop here."  Leading me to leaving the majority of "treeobj errors" as fatal.  

But lemme go through and consider it again.

Reply:
Design wise, I think this tool should try to report (and eventually unlink/fix)) all the damaged keys, however unlikely the nature of the damage is.